sources/ghostdog/src/main.rs

```diff
@@ -160,6 +175,51 @@ fn is_neuron_attached() -> Result<()> {
     }
 }
 
+fn is_infiniband_attached() -> Result<()> {
+    // confirm device has `SW_MNG` in the VPD file
+    let infiniband_sys_devices =
+        fs::read_dir("/sys/class/infiniband/").context(error::InfinibandSysDevicesSnafu)?;
+
+    for device in infiniband_sys_devices {
+        let device = device.context(error::InfinibandSysDevicesSnafu)?;
+        let device_name: OsString = device.file_name().clone();
+        let vpd_file = Path::new(&device.path()).join("device").join("vpd");
+        if vpd_file.exists() {
+            let vpd_file_bytes: &[u8] =
+                &std::fs::read(vpd_file.as_path()).context(error::ReadFileSnafu {
+                    path: vpd_file.as_path(),
+                })?;
+            let vpd_file_string = String::from_utf8_lossy(vpd_file_bytes);
+            if !vpd_file_string.contains("SW_MNG") {
+                continue;
+            }
```

Comment:
Might be worth isolating this so you can write a unit test against a real `vpd` file.

---

I've never seen a `vpd` file. If it's small, you can likely ignore this feedback. The `utf8` conversion can be skipped here though, since we don't actually care about the encoding and our "needle" is ASCII.

```suggestion
            let vpd_file_bytes: &[u8] =
                &std::fs::read(vpd_file.as_path()).context(error::ReadFileSnafu {
                    path: vpd_file.as_path(),
                })?;
            let needle = "SW_MNG".as_bytes();
            if vpd_file_bytes.windows(needle.len())
                .any(|window| window == needle) {
                continue;
            }
```

Reply:
I was trying to find a way to get this data out of what is a very small (binary) file and this works, but a cleaner way would be to parse the `lspci` output based upon the slot. Given a device `/sys/class/infiniband/ibp115s0f0` with a `device` link of `../../../0000:73:00.0` (`readlink device`), you can use that to get just data about that device:

<details>

```
73:00.0 Infiniband controller: Mellanox Technologies MT2910 Family [ConnectX-7]
	Subsystem: Mellanox Technologies Device 0087
	Physical Slot: 107-1
	Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0
	Interrupt: pin A routed to IRQ 254
	NUMA node: 1
	Region 0: Memory at a8000000 (64-bit, prefetchable) [size=32M]
	Capabilities: [60] Express (v2) Endpoint, IntMsgNum 0
		DevCap:	MaxPayload 512 bytes, PhantFunc 0, Latency L0s unlimited, L1 unlimited
			ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset+ SlotPowerLimit 25W TEE-IO-
		DevCtl:	CorrErr+ NonFatalErr+ FatalErr+ UnsupReq-
			RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+ FLReset-
			MaxPayload 128 bytes, MaxReadReq 4096 bytes
		DevSta:	CorrErr+ NonFatalErr- FatalErr- UnsupReq+ AuxPwr- TransPend-
		LnkCap:	Port #0, Speed 16GT/s, Width x2, ASPM not supported
			ClockPM- Surprise- LLActRep- BwNot- ASPMOptComp+
		LnkCtl:	ASPM Disabled; RCB 64 bytes, LnkDisable- CommClk-
			ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
		LnkSta:	Speed 16GT/s, Width x2
			TrErr- Train- SlotClk+ DLActive- BWMgmt- ABWMgmt-
		DevCap2: Completion Timeout: Range ABC, TimeoutDis+ NROPrPrP- LTR-
			 10BitTagComp+ 10BitTagReq+ OBFF Not Supported, ExtFmt- EETLPPrefix-
			 EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
			 FRS- TPHComp- ExtTPHComp-
			 AtomicOpsCap: 32bit+ 64bit+ 128bitCAS+
		DevCtl2: Completion Timeout: 260ms to 900ms, TimeoutDis-
			 AtomicOpsCtl: ReqEn+
			 IDOReq- IDOCompl- LTR- EmergencyPowerReductionReq-
			 10BitTagReq+ OBFF Disabled, EETLPPrefixBlk-
		LnkCap2: Supported Link Speeds: 2.5-16GT/s, Crosslink- Retimer+ 2Retimers+ DRS-
		LnkCtl2: Target Link Speed: 16GT/s, EnterCompliance- SpeedDis-
			 Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
			 Compliance Preset/De-emphasis: -6dB de-emphasis, 0dB preshoot
		LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete+ EqualizationPhase1+
			 EqualizationPhase2+ EqualizationPhase3+ LinkEqualizationRequest-
			 Retimer- 2Retimers- CrosslinkRes: unsupported
	Capabilities: [48] Vital Product Data
		Product Name: Nvidia ConnectX-7 mezz internal for Nvidia Umbriel system
		Read-only fields:
			[PN] Part number: 692-9X760-00SE-S00
			[EC] Engineering changes: A4
			[V2] Vendor specific: 692-9X760-00SE-S00
			[SN] Serial number: MT2451601XLH
			[V3] Vendor specific: 502d1f0053bfef118000e09d733f3bf8
			[VA] Vendor specific: MLX:MN=MLNX:CSKU=V2:UUID=V3:PCI=V0:SMDL=SW_MNG:MODL=C7010Z
			[V0] Vendor specific: PCIeGen4 x2
			[VU] Vendor specific: MT2451601XLHMLNXS0D0F0
			[RV] Reserved: checksum good, 2 byte(s) reserved
		End
	Capabilities: [9c] MSI-X: Enable+ Count=64 Masked-
		Vector table: BAR=0 offset=00002000
		PBA: BAR=0 offset=00003000
	Capabilities: [c0] Vendor Specific Information: Len=18 <?>
	Capabilities: [40] Power Management version 3
		Flags: PMEClk- DSI- D1- D2- AuxCurrent=375mA PME(D0-,D1-,D2-,D3hot-,D3cold+)
		Status: D0 NoSoftRst+ PME-Enable- DSel=0 DScale=0 PME-
	Kernel driver in use: mlx5_core
```

</details>

Where there this data is easier to find and in text:
```
[VA] Vendor specific: MLX:MN=MLNX:CSKU=V2:UUID=V3:PCI=V0:SMDL=SW_MNG:MODL=C7010Z
```
I think the reading of the output might be cleaner than this, but it *is* a tiny file, even if a little strange to search a string in a binary file (thus the strange parsing I'm doing here).

Reply:
We discussed this aloud and decided that there is a slim but imagineable chance that the binary contents of the `vpd` file could trick `to_string_lossy()` into incorrectly escaping the relevant ASCII string due to misidentifying binary contents of the file as variable content-length code points.

It's probably a good idea to adopt the "bytestring search" window approach.

Reply:
This new approach works great! Thanks for the tip!